### PR TITLE
[DET-2628] docs: Add a topic guide about the Native API

### PR DESCRIPTION
**Pull request by yoavz**

### CODE REVIEW

File: docs/topic-guides/native-api.txt [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+<https://docs.python.org/3/library/runpy.html#runpy.run_path>`_. In the trial
+runner environment, ``init`` will return a :py:class:`determined.NativeContext`
+object that holds information specific to that trial (e.g. hyperparameter
+choices) and continue executing. Once the script hits the training loop
+function (in this case, ``tf.keras.Models.fit``), Determined will launch into
+the managed training loop.
+
+.. warning::
+
+    Any user code that happens occurs after the training loop has started will
```
Comment (vishnu2kmohan) [inline docs/topic-guides/native-api.txt:53]: remove `happens`?

File: docs/topic-guides/native-api.txt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    never be executed!
+
+High-Level Training Loop Support
+================================
+
+In order to execute a fully custom Python script, Determined requires that the
```
Comment (vishnu2kmohan) [inline docs/topic-guides/native-api.txt:59]: Nit: remove `fully`

File: docs/topic-guides/native-api.txt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+user script contains a high-level training loop that it can "hijack".
```
Comment (vishnu2kmohan) [inline docs/topic-guides/native-api.txt:60]: `"hijack"` scares me (and possibly, our target user) a little. 
I'd suggest that we replace it with `intercept`, which feels a little more innocuous.

File: docs/topic-guides/native-api.txt [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+.. _topic-guide_native_trial-api:
+
+Integration with the Trial API
+==============================
+
+The Native API can also be used to submit experiments using the
+:ref:`pytorch-trial`, :ref:`tf-keras-trial`, or :ref:`estimator-trial` APIs. To
+do so, write a script that passes our ``Trial`` class definition and
+configuration to :py:func:`determined.experimental.create()` and execute it in
```
Comment (vishnu2kmohan) [inline docs/topic-guides/native-api.txt:78]: Just a comment but I had to look twice at `determined.experimental.create()` because I was expecting to read `determined.experiment.create()`

Once the Native API is ready to be marked stable (and therefore removed from the `experimental` namespace), `determined.create()` feels like it _ought_ to be capable of launching more than just experiments (e.g., Commands, Shells, Notebooks and TensorBoards) which may disappoint the user, in the future.

File: docs/topic-guides/native-api.txt (hunk 3 above)
Comment (yoavz) [reply docs/topic-guides/native-api.txt]: replaced

File: docs/topic-guides/native-api.txt [hunk 5]
```diff
@@ -0,0 +1,88 @@
+.. _native-api:
+
+Native API (experimental)
+=========================
+
+.. warning:
+
+    The Native API is currently an experimental API.
+
+The Native API allows developers to seamlessly move between training in local
```
Comment (shiyuann) [inline docs/topic-guides/native-api.txt:10]: The first sentence seems a bit ambiguous to me in what the main problem Native API wants to solve. When you say "training in local Python script", do you mean writing code for training in your local machine or literally training in local Python scripts? If it's the latter, it has the same meaning as the local mode does.

What do you think of the following sentences as the first paragraph?

Native API provides a very easy-to-use API for us to easily making use of Determined's capability to train multiple models on multiple machines. It supports Tensorflow Keras and Estimator with idiomatic framework patterns. A typical use case is that before using Determined we have training code written to integrate with the native frameworks and we want to port the model to the Determined API. Using the Native API, we only need minimum effort to do this.

File: docs/topic-guides/native-api.txt [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+.. TODO: Add a link to show what an experiment in local development mode looks like
+
+The diagram above demonstrates the flow of execution when an experiment is
+created in :py:class:`determined.experimental.Mode.CLUSTER
+<determined.experimental.Mode>` mode.
+
+The Python script will first be executed by us in the *User environment*, such
+as a Python virtualenv or a Jupyter notebook where the ``determined`` Python
+package is installed. When the code is executed, ``init()`` will package up all
+the files in the provided context directory, and create an experiment by
```
Comment (shiyuann) [inline docs/topic-guides/native-api.txt:35]: > create an experiment by submitting them along with the :ref:`experiment configuration <terminology-concepts>`

should be "by submitting the context directory"?

File: docs/topic-guides/native-api.txt [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+submitting them along with the :ref:`experiment configuration
+<terminology-concepts>` in a network request to the Determined master. Note
+that any code that comes after the ``init()`` call is *not* executed in our
```
Comment (shiyuann) [inline docs/topic-guides/native-api.txt:38]: > Note that any code that comes after the ``init()`` call is *not* executed in our local environment

"in our local environment" -> "when submitting the code"

File: docs/topic-guides/native-api.txt [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+local environment -- in this case building and compiling the model is only done
```
Comment (shiyuann) [inline docs/topic-guides/native-api.txt:39]: "building and compiling the model" -> "the code after the ``init()`` call that typically involves defining the model and training"

File: docs/topic-guides/native-api.txt [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+on the Determined cluster.
+
+Once the Determined master has initialized a *Trial Runner environment*, the
+user script is re-executed from the beginning using `runpy.run_path
```
Comment (shiyuann) [inline docs/topic-guides/native-api.txt:43]: "user script is re-executed from the beginning using runpy.run_path" -> "Determined loads the model and other training-related code by using `runpy.run_path() <https://docs.python.org/3/library/runpy.html#runpy.run_path>`_."

File: docs/topic-guides/native-api.txt (hunk 6 above)
Comment (yoavz) [reply docs/topic-guides/native-api.txt:35]: updated wording

File: docs/topic-guides/native-api.txt (hunk 7 above)
Comment (yoavz) [reply docs/topic-guides/native-api.txt:38]: it is executed when submitting the code though, just in the remote env. I think "in our local environment" is clearer in this case.

File: docs/topic-guides/native-api.txt (hunk 8 above)
Comment (yoavz) [reply docs/topic-guides/native-api.txt:39]: Reworded it a bit -- I still wanted to keep the "is only done on the Determined cluster" part but I added your suggestion to earlier in the sentence.

File: docs/topic-guides/native-api.txt (hunk 9 above)
Comment (yoavz) [reply docs/topic-guides/native-api.txt:43]: I prefer being explicit that the user script is re-executed from the beginning. "Determined loads the model and other training-related code" leaves it more ambiguous imo.

File: docs/topic-guides/native-api.txt (hunk 5 above)
Comment (yoavz) [reply docs/topic-guides/native-api.txt:10]: (I copied this paragraph straight from the Native API tutorial)

I modified "training in local Python scripts" to "training in a local development environment" on both to make that clearer. Apart from that, I think the paragraph as it stands is clearer than the one you've proposed -- I guess it's a matter of style at the end of the day...

Comment (shiyuann) [reply docs/topic-guides/native-api.txt]: Do you mean training in a local development environment with Determined or without Determined?

File: docs/topic-guides/native-api.txt (hunk 8 above)
Comment (shiyuann) [reply docs/topic-guides/native-api.txt]: yeah. "is only done on the Determined cluster" is good. I guess what I was proposing is that in Estimator, there is no explicit model building and compiling.

File: docs/topic-guides/native-api.txt (hunk 9 above)
Comment (shiyuann) [reply docs/topic-guides/native-api.txt:43]: got it. My point is the user script is executed twice but for different purposes. But I can see your point

File: docs/topic-guides/native-api.txt [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+idiomatic framework patterns, reducing (or eliminating) the effort to port
+model code for use with Determined.
+
+In this guide, we'll cover what happens under the hood when a user submits an
+experiment to Determined using the Native API. This topic guide is meant as
+deep-dive for advanced users -- for a tutorial on getting started with the
+Native API, see :ref:`tutorials_native-api`.
+
+Life of a Native API Experiment
+===============================
```
[the remark this answers is not in the document]
Comment (shiyuann) [reply docs/topic-guides/native-api.txt:22]: got it. But they are rendered at the same level as other titles. 

The list is as below:

```
Optimizing Distributed and Parallel Training
Reproducibility
Scheduling
Terminology and Concepts
User Interfaces
YAML Topic Guide
Native API (experimental)
Life of a Native API Experiment
High-Level Training Loop Support
Integration with the Trial API
```

Seems weird if you click in "Integration with the Trial API" since it's at the first level and find it's actually about Native API. "Integration with the Trial API" alone seems to lose the context of using Native API.

File: docs/topic-guides/native-api.txt (hunk 7 above)
Comment (shiyuann) [reply docs/topic-guides/native-api.txt]: I see the structure of this paragraph. BTW, it seems it should be "in the User environment" to be consistent with other uses in this paragraph.

File: docs/topic-guides/native-api.txt (hunk 5 above)
Comment (yoavz) [reply docs/topic-guides/native-api.txt]: without Determined

Comment (yoavz) [reply docs/topic-guides/native-api.txt]: "local development environment" implies not controlled by Determined

File: docs/topic-guides/native-api.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (yoavz) [reply docs/topic-guides/native-api.txt:22]: My bad, you're totally right. I meant to make them `----`, indexed on the topic guide itself but not in the topic guide `index`. modified.

File: docs/topic-guides/native-api.txt (hunk 7 above)
Comment (yoavz) [reply docs/topic-guides/native-api.txt]: updated

File: docs/topic-guides/native-api.txt (hunk 8 above)
Comment (yoavz) [reply docs/topic-guides/native-api.txt]: Yep, I added a `tf.keras` note before this phrase.

**Comment by yoavz** [comment]
Here is the link for future reference: https://docs.google.com/drawings/d/1Uq5Uq4kdY7gp2IXXb_v-jVIt-i771vexvrAd5y13hCs/edit

**Comment by yoavz** [comment]
I'm not sure I'd encode a google drive link in the doc source code, but I have placed it in a relevant gdrive location any Determined employees can access and modify.

**Comment by yoavz** [comment]
cla-bot has add the signed label here, even though the github check is pending. i'm going to assume the github check is broken and rebase / merge.
